## Fraud Detection - Preprocessing
Cleaning data and engineering features based on EDA findings

In [1]:
# Load libraries and raw dataset
import pandas as pd
import numpy as np

df = pd.read_csv("../data/fraud_transactions.csv")
print(df.shape)

(6362620, 11)


In [2]:
# Drop columns not useful for modeling: names are just IDs, isFlaggedFraud is a weak naive rule
df = df.drop(columns=['nameOrig', 'nameDest', 'isFlaggedFraud'])
df.shape

(6362620, 8)

In [3]:
# Feature: difference between old and new origin balance (fraud tends to drain accounts)
df['balance_diff_orig'] = df['oldbalanceOrg'] - df['newbalanceOrig']
df[['oldbalanceOrg', 'newbalanceOrig', 'balance_diff_orig']].head()

,oldbalanceOrg,newbalanceOrig,balance_diff_orig
0,170136.0,160296.36,9839.64
1,21249.0,19384.72,1864.28
2,181.0,0.00,181.00
3,181.0,0.00,181.00
4,41554.0,29885.86,11668.14


### Engineered Feature: balance_diff_orig
- `oldbalanceOrg - newbalanceOrig` — captures how much left the origin account
- EDA showed fraud tends to drain accounts to $0, this quantifies that pattern

In [4]:
# One-hot encode transaction type
df = pd.get_dummies(df, columns=['type'], drop_first=True)
df.columns.tolist()

['step',
 'amount',
 'oldbalanceOrg',
 'newbalanceOrig',
 'oldbalanceDest',
 'newbalanceDest',
 'isFraud',
 'balance_diff_orig',
 'type_CASH_OUT',
 'type_DEBIT',
 'type_PAYMENT',
 'type_TRANSFER']

### Encoded Type
- One-hot encoded `type` with drop_first=True (CASH_IN as baseline)
- EDA showed fraud only occurs in TRANSFER/CASH_OUT — this lets the model use that signal directly

In [5]:
# Save cleaned dataset
df.to_csv("../data/fraud_transactions_clean.csv", index=False)
print("Saved:", df.shape)

Saved: (6362620, 12)


### Saved Clean Dataset
- Saved to `data/fraud_transactions_clean.csv` (not pushed to GitHub per assignment instructions)
- Shape: (6362620, 12)